[Tablet Orders] Use recalculate button for selection in side-by-side mode

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #11908
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR moves our side-by-side order creation experience to use a `Recalculate` button, instead of the `immediate` sync approach. Users must tap `Recalculate` in side-by-side mode to sync the order changes before they can continue with creating the order.

It’s added as temporary mitigation for selection sync difficulties.

We’ll go back to immediate in the near future, so that code is left in for ease later.

Note that this PR also moves the responsibility for triggering sync to the EditableOrderViewModel, via the previously existing closures such as `onProductSelectionStateChanged` or `onAllSelectionsCleared`. This seems like less of a side-effect than having the Product Selector call `onMultipleSelectionCompleted` at appropriate times.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

1. Enable the `sideBySideViewForOrderForm` feature flag
2. Launch the app on an iPad or large iPhone in landscape
3. Navigate to the `Orders` tab, and tap `+` to open the order form
4. Add products to the order
5. Observe that the button shows `Recalculate`
6. Tap `Recalculate` – observe that the order summary is updated, and the completion button becomes `Collect Payment`.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/woocommerce/woocommerce-ios/assets/2472348/df7e3c2d-9a78-4844-a598-f18575edec83


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
11908 handle sync decisions in main view model

joshheald committed:
11908 Use recalculate button in side by side mode

This button is instead of the `immediate` sync approach. It’s added as temporary mitigation for selection sync difficulties.

We’ll go back to immediate in the near future, so that code is left in for ease later.

joshheald committed:
Merge branch 'trunk' into issue/11908-use-recalculate-button-for-selection

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift
#	WooCommerce/WooCommerceTests/Mocks/MockFeatureFlagService.swift
#	WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift

joshheald committed:
11908 fix merge mistakes

joshheald committed:
11908 Ensure that we keep selections on mode change

Previously we would lose unsynced selection changes, particularly when moving from side-by-side to modal-on-modal.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/188">17.6</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
11908 Update nav button for order creation

joshheald committed:
11908 Remove outdated test

As of 61a850450abab205d1464942bc012c945ed91d86, we handle this logic in the EditableOrderViewModel, so this test should have previously been removed.

joshheald committed:
11908 Fix merge regression

iamgabrielma commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/EditableOrderViewModelTests.swift:
```diff
@@ -289,6 +289,75 @@ final class EditableOrderViewModelTests: XCTestCase {
+
+        // Then
+        switch viewModel.doneButtonType {
+        case .recalculate:
+            // Success – we just don't care about the `loading` parameter
```
TIL we don't need to pass any associated value for testing, and we can just pattern-match it, then make the test succeed when we don't land on the XCTFail case.

joshheald replied:
I still don't love it; the implicit pass isn't that nice to read. It's good to ignore the associated value though (without writing an `==` implementation that ignores it – I ran across that in a test the other day, we'd added it in the past and so my test which _did_ care about the associated value was passing even when it shouldn't have!)

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift:
```diff
@@ -181,7 +181,7 @@ struct OrderForm: View {
         orderFormSummary(presentProductSelector)
             .onAppear {
-                viewModel.syncChangesImmediately = presentationStyle == .sideBySide
+                viewModel.selectionSyncApproach = presentationStyle == .sideBySide ? .onRecalculateButtonTap : .onSelectorButtonTap
```
Would it be more readable if we move this bit to a computed variable outside of the view's body?
```    
private var selectionSyncApproach: EditableOrderViewModel.OrderItemSelectionSyncApproach {
        presentationStyle == .sideBySide ? .onRecalculateButtonTap : .onSelectorButtonTap
    }

    var body: some View {
        orderFormSummary(presentProductSelector)
            .onAppear {
                viewModel.selectionSyncApproach = selectionSyncApproach
            }
    }
```

joshheald replied:
I'll take a look, but probably still on the view somewhere. It's about converting `presentationStyle` to `syncApproach`, and it feels like it fits better in the OrderForm view, which covers both the relevant views.

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -1227,24 +1250,63 @@ private extension EditableOrderViewModel {
 
+    /// Calculates what Call to Action button should be shown depending on our internal state.
+    ///
+    func configureDoneButton() {
```
Really nice way to setup both the done button and navigation items  💯 I need to copy this approach in those places where the rendering depends on multiple computed variables.

joshheald replied:
I just updated and copied it, tbh, but it is a nice contained use of combine which isn't _too_ convoluted. Breaking it down into separate publishers for the `isPerformingNetworkRequest` concept helped.

iamgabrielma reviewed (approved):
Works great! :shipit: 

At some point during testing I managed to break the variations selected state somehow (none selected, but on tapping recalculate added them to the order) but I haven't been able to reproduce it. 

Definitely an improvement UX-wise, as well as moving the sync responsibility up the chain from the selector view to the order creation 💯

joshheald commented:
> Works great! :shipit:

Thanks for the review!

> At some point during testing I managed to break the variations selected state somehow (none selected, but on tapping recalculate added them to the order) but I haven't been able to reproduce it.

Yep, this is totally possible. I meant to mention it in the description but forgot, sorry!

It predates this PR, so I added a separate ticket for it: #12135

joshheald committed:
11908 fix whitespace

joshheald committed:
11908 tidy up selection sync approach update
